File: src/09-challenge/solution.py
from typing import Literal, Dict, Tuple, Callable

Move = Literal["L", "R", "U", "D"]
Result = Literal["fail", "crash", "success"]

MoveActions = Dict[Move, Callable[[int, int], Tuple[int, int]]]


def fn(board: str, moves: str) -> Result:
    _board = list(list(row) for row in board.strip().split("\n"))

    robot_x: int = -1
    robot_y: int = -1

    for y, row in enumerate(_board):
        for x, col in enumerate(row):
            if col == "@":
                robot_x = x
                robot_y = y

    move_actions: MoveActions = {
        "L": lambda x, y: (x - 1, y),
        "R": lambda x, y: (x + 1, y),
        "U": lambda x, y: (x, y - 1),
        "D": lambda x, y: (x, y + 1),
    }

    for move in moves:
        next_robot_coords = move_actions[move](robot_x, robot_y)  # type: ignore

        if (
            next_robot_coords[0] < 0
            or next_robot_coords[0] >= len(_board[0])
            or next_robot_coords[1] < 0
            or next_robot_coords[1] >= len(_board)
            or _board[next_robot_coords[1]][next_robot_coords[0]] == "#"
        ):
            return "crash"

        _board[robot_y][robot_x] = "."

        robot_x = next_robot_coords[0]
        robot_y = next_robot_coords[1]

        if _board[robot_y][robot_x] == "*":
            return "success"

        _board[robot_y][robot_x] = "@"

    return "fail"

File: src/09-challenge/solution.js
/**
 * @typedef {string} Board
 * @typedef {'L' | 'R' | 'U' | 'D'} Move
 * @typedef {'fail' | 'crash' | 'success'} Result
 * @typedef {Record<Move | 'undefined', () => [number, number]>} MoveActions
 */

/**
 * @param {string} board - Represent the board situation
 * @param {string} moves - Movement direction
 * @returns {Result}
 */
export default function fn(board, moves) {
	const _board = board
		.trim()
		.split('\n')
		.map((row) => row.split(''))

	let robotY = _board.findIndex((row) => row.includes('@'))
	let robotX = _board[robotY].indexOf('@')

	/** @type {MoveActions} */
	const moveActions = {
		L: () => [robotX - 1, robotY],
		R: () => [robotX + 1, robotY],
		U: () => [robotX, robotY - 1],
		D: () => [robotX, robotY + 1],
		undefined: () => [-1, -1],
	}

	/** @type {Result} */
	let result = 'fail'

	for (let i = 0; i < moves.length && result === 'fail'; i++) {
		const move = moves[i]
		const nextRobotI = moveActions[move]()

		if (
			nextRobotI[0] < 0 ||
			nextRobotI[0] >= _board[0].length ||
			nextRobotI[1] < 0 ||
			nextRobotI[1] >= _board.length ||
			_board[nextRobotI[1]][nextRobotI[0]] === '#'
		) {
			result = 'crash'
			continue
		}

		_board[robotY][robotX] = '.'

		robotX = nextRobotI[0]
		robotY = nextRobotI[1]

		if (_board[robotY][robotX] === '*') result = 'success'

		_board[robotY][robotX] = '@'
	}

	return result
}
